cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+    checks ["Submission not readable: " <> pack afp]
+  unlessM (lift $ Dir.doesDirectoryExist root) $
+    checks ["Registry root doesn't exist: " <> pack afp]
+
+  let (dir', file) = Path.splitFileName afp
+      dir = Path.dropTrailingPathSeparator dir'
+      (parentParentDir, parentDir) = Path.splitFileName dir
+
+  checks $
+    ["Submission not in 'registry' directory"
+    | parentDir /= "registry"] <>
+    ["Submission not properly relative the registry root"
+    | Path.dropTrailingPathSeparator root
+      /= Path.dropTrailingPathSeparator parentParentDir]
+
+  let (id, ext) = Path.splitExtension file
+      sigPath = dir </> id <.> "sig"
+
+  unlessM (lift $ Dir.doesFileExist sigPath) $
+    checks ["Missing signature file: " <> pack sigPath]
+  unlessM (lift $ Dir.readable <$> Dir.getPermissions sigPath) $
+    checks ["Signature file not readable: " <> pack sigPath]
+  checks $
+    ["Submission filename doesn't have .json extension: " <> pack file
+    | ext /= ".json" ]
+
+  pubkey <- case validatePublicKey (pack id) of
+    Left e -> err $
+      ["Submission filename (actually public key) invalid: " <> e]
+    Right x -> pure x
+
+  contents <- lift $ BSL.fromStrict <$> BS.readFile afp
+  case AE.eitherDecode contents of
+    Left e -> err $ ["Submission content:\n" <> pack e]
+    Right entry -> do
+
+      checks $
+        [ pack $ "Internal 'id' doesn't match filename: "
+          <> show (unId $ sId entry)  <> " vs. " <> show (unId pubkey)
+        | sId entry /= pubkey ]
+
+      -- All cheap checks done, let's do the expensive one now.
+      dupTickers <- withExceptT (PoolRegistryValidationErrors fp . (:[]) .
+                               ("While reading all registry entries: "<>)) $
+        duplicateTickers dir
+
+      checks $
+        [ "Duplicate tickers: " <>
+          pack (unpack $ Text.intercalate ", " $ unTicker <$> dupTickers)
+        | not (null dupTickers) ]
+
+      -- All valid.
+      pure entry
+
+ where
+   checks :: [Text] -> ExceptT CliError IO ()
+   checks [] = ExceptT . pure . Right $ ()
+   checks es = err es
+
+   err :: [Text] -> ExceptT CliError IO a
+   err es = ExceptT . pure . Left  $ PoolRegistryValidationErrors fp es
+
+   duplicateTickers :: FilePath -> ExceptT Text IO [Ticker]
+   duplicateTickers dir = do
+     paths <- ((dir </>) <$>) <$> (lift $ Dir.listDirectory dir)
+     -- We maintain the invariant that 'registry' contains only files.
+     tickers <- withExceptT pack $ ExceptT $
+       sequence <$> traverse entryTicker (selectEntries paths)
+
+     pure . Set.toList $ duplicates tickers
+     where
+       selectEntries :: [FilePath] -> [FilePath]
+       selectEntries paths =
+         flip filter paths $ (==".json") . Path.takeExtension
+
+       entryTicker :: FilePath -> IO (Either String Ticker)
+       entryTicker f =
+         (bimap (("While parsing '"<>f<>"': ")<>) sTicker) . AE.eitherDecode . BSL.fromStrict
+           <$> BS.readFile f
```

Comment:
The logic here seems a bit backward to me :thinking: 
What about: 

- Get the list of all existing tickers

```hs 
    listTickers :: FilePath -> ExceptT Test IO [Ticker]
```

- Check whether the new ticker is part of that list 

```hs
tickers <- listTickers dir; 
when (myTicker `elem` tickers) $ 
    throwE (validationError "ticker already known")
```

Reply:
That would assume that there are no other duplicate tickers -- which is a flaky assumption I'd rather avoid.

Reply:
I am not sure to follow you on this? Even if there are duplicates ticker already in the list, it will still work fine, no?

Reply:
You are suggesting to only validate the newly-added ticker for duplication -- that wouldn't find other duplicates, would it?

Reply:
Indeed.

Reply:
So as I said, I intended to look for other duplicates too.

We already paid to the heavy cost of going to the kernel to find the files and open them and read them -- the extra check that there is no validation in the pre-existing set is peanuts, and it adds value -- I believe.